test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java

```diff
@@ -532,4 +536,35 @@ protected static boolean aliasExists(String index, String alias) throws IOExcept
         assertNotNull(responseEntity);
         return responseEntity;
     }
+
+    /**
+     * Sniff the cluster for host metadata if it hasn't already been sniffed. This isn't the
+     * same thing as using the {@link Sniffer} because:
+     * <ul>
+     * <li>It doesn't replace the hosts that that {@link #client} communicates with
+     * <li>It only runs once
+     * </ul>
+     */
+    protected void sniffHostMetadata(RestClient client) throws IOException {
+        if (HostMetadata.EMPTY_RESOLVER != client.getHostMetadataResolver()) {
+            // Already added a resolver
+            return;
+        }
+        // No resolver, sniff one time and resolve metadata against the results
+        ElasticsearchHostsSniffer.Scheme scheme;
+        switch (getProtocol()) {
+        case "http":
+            scheme = ElasticsearchHostsSniffer.Scheme.HTTP;
+            break;
+        case "https":
+            scheme = ElasticsearchHostsSniffer.Scheme.HTTPS;
+            break;
+        default:
+            throw new UnsupportedOperationException("unknown protocol [" + getProtocol() + "]");
+        }
+        ElasticsearchHostsSniffer sniffer = new ElasticsearchHostsSniffer(
+            adminClient, ElasticsearchHostsSniffer.DEFAULT_SNIFF_REQUEST_TIMEOUT, scheme);
+        Map<HttpHost, HostMetadata> meta = sniffer.sniffHosts().hostMetadata();
+        client.setHosts(clusterHosts, meta::get);
```

Comment:
They, uh, have so far!

Now that I think about it I think I should try and be a little more careful here and assert that all the hosts come back. The test would fail in a weird way if I don't.